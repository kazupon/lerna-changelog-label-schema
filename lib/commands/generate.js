"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const debug_1 = require("debug");
const debug = debug_1.debug('lerna-changelog-label-schema:commands:generate');
const generate = async (options) => {
    const { preset, output } = options;
    debug(`preset= ${preset}, output=${output}`);
    if (!(preset === 'default' || preset === 'full')) {
        return false;
    }
    const schema = await Promise.resolve().then(() => __importStar(require(`../schema/${preset}`)));
    const labels = schema.labels;
    const meta = schema.meta;
    debug('labels', labels);
    debug('meta', meta);
    const githubLabels = labels.map((label, index) => {
        const m = meta[index];
        if (m.type !== label.type) {
            throw new Error(`Invalid generating: label=${label.type}, meta=${m.type}, index=${index}`);
        }
        const ghLabel = {
            name: `Type: ${label.type}`,
            color: m.color
        };
        if (label.description) {
            ghLabel.description = label.description;
        }
        if (m.aliases) {
            ghLabel.aliases = m.aliases;
        }
        return ghLabel;
    });
    debug('githubLabels', githubLabels);
    let ret = true;
    if (output) {
        try {
            await utils_1.writeJSON(output, githubLabels);
        }
        catch (e) {
            console.error(e.message);
            ret = false;
        }
    }
    else {
        console.log(JSON.stringify(githubLabels, null, 2));
    }
    return ret;
};
exports.default = generate;
