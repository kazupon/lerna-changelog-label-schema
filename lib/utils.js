"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
async function readJSON(path) {
    return await Promise.resolve().then(() => __importStar(require(path)));
}
exports.readJSON = readJSON;
async function writeJSON(path, json) {
    const data = JSON.stringify(json, null, 2);
    return new Promise((resolve, reject) => {
        fs.writeFile(path, data, err => {
            err ? reject(err) : resolve();
        });
    });
}
exports.writeJSON = writeJSON;
