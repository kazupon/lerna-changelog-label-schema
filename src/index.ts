import * as _default from './schema/default'
import * as full from './schema/full'

export const schema = {
  default: _default,
  full
}
