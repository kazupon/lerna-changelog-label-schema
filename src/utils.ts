import * as fs from 'fs'

export async function readJSON (path: string) {
  return await import(path) as JSON
}

export async function writeJSON (path: string, json: any) {
  const data = JSON.stringify(json, null, 2)
  return new Promise((resolve, reject) => {
    fs.writeFile(path, data, err => {
      err ? reject(err) : resolve()
    })
  })
}
