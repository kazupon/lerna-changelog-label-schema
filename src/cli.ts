import meow from 'meow'
import { define, generate } from './commands/index'

import { debug as Debug } from 'debug'
const debug = Debug('lerna-changelog-label-schema:cli')

export async function run (argv: string[] = []) {
  const options: meow.Options = {
    flags: {
      preset: {
        type: 'string',
        alias: 'p',
        default: 'default'
      },
      output: {
        type: 'string',
        alias: 'o'
      }
    }
  }

  if (argv.length > 0) {
    options.argv = argv
  }

  const cli = meow(`
  Usage
    $ lerna-changelog-label-schema <command> [options]
  
  Commands:
    define, d      define lerna-changelog labels in your package.json 
    generate, g    generate lerna-changelog labels for GitHub labels, output for stdout
    version, v     show the version

  Options:
    --preset, -p   Label schema preset option, for define and generate command. 'default' or 'full', default: 'default'
    --output, -o   Output option, for generate command. if specified, stdout is suppressed.
  
  Examples:
    $ lerna-changelog-label-schema define --preset=full
    $ lerna-changelog-label-schema generate --preset=default
  `, options)

  const command = cli.input.pop() || ''
  debug('command:', command)
  const { preset, output } = cli.flags
  debug('flags:', cli.flags)

  if (['define', 'd'].includes(command)) { // define command
    await define({ preset, cwd: process.cwd() })
  } else if (['generate', 'g'].includes(command)) { // generate command
    await generate({ preset, output })
  } else if (['version', 'v'].includes(command)) { // version command
    console.log(cli.pkg.version)
  } else { // help!
    console.log(cli.help)
  }

  return cli
}
