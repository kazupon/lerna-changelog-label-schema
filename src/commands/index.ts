import define from './define'
import generate from './generate'

export {
  define,
  generate
}
