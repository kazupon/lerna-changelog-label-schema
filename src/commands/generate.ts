import { writeJSON } from '../utils'
import {
  GitHubLabelMeta,
  LernaChangelogLabel,
  GitHubLabel,
  CommandHandler,
  CommandOption
} from '../../types'

import { debug as Debug } from 'debug'
const debug = Debug('lerna-changelog-label-schema:commands:generate')

const generate: CommandHandler = async (options: CommandOption) => {
  const { preset, output } = options
  debug(`preset= ${preset}, output=${output}`)

  if (!(preset === 'default' || preset === 'full')) {
    return false
  }

  const schema = await import(`../schema/${preset}`)
  const labels = schema.labels as LernaChangelogLabel[]
  const meta = schema.meta as GitHubLabelMeta[]
  debug('labels', labels)
  debug('meta', meta)

  const githubLabels = labels.map((label, index) => {
    const m = meta[index]
    if (m.type !== label.type) {
      throw new Error(`Invalid generating: label=${label.type}, meta=${m.type}, index=${index}`)
    }
    const ghLabel: GitHubLabel = {
      name: `Type: ${label.type}`,
      color: m.color
    }
    if (label.description) {
      ghLabel.description = label.description
    }
    if (m.aliases) {
      ghLabel.aliases = m.aliases
    }
    return ghLabel
  }) as GitHubLabel[]
  debug('githubLabels', githubLabels)

  let ret = true
  if (output) {
    try {
      await writeJSON(output as string, githubLabels)
    } catch (e) {
      console.error(e.message)
      ret = false
    }
  } else {
    console.log(JSON.stringify(githubLabels, null, 2))
  }

  return ret
}

export default generate
