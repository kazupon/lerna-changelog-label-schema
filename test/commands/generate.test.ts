import * as path from 'path'
import generate from '../../src/commands/generate'

// ----------
// mocks

jest.mock('../../src/utils', () => ({
  ...jest.requireActual('../../src/utils'),
  writeJSON: jest.fn()
}))
import * as utils from '../../src/utils'

// -------------------
// setup/teadown hooks

let spyLog
let spyError
beforeEach(() => {
  spyLog = jest.spyOn(global.console, 'log')
  spyError = jest.spyOn(global.console, 'error')
})

afterEach(() => {
  spyError.mockRestore()
  spyLog.mockRestore()
  jest.clearAllMocks()
})

// ----------
// tests

test('preset: default', async () => {
  // run
  const ret = await generate({ preset: 'default' })

  // verify
  expect(ret).toBe(true)
  expect(spyLog.mock.calls[0][0]).toMatchSnapshot()
})

test('preset: full', async () => {
  // run
  const ret = await generate({ preset: 'full' })

  // verify
  expect(ret).toBe(true)
  expect(spyLog.mock.calls[0][0]).toMatchSnapshot()
})

test('preset: other', async () => {
  // run
  const ret = await generate({ preset: 'foo' })

  // verify
  expect(ret).toBe(false)
})

test('output: success', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.writeJSON.mockImplementation((path, json) => (Promise.resolve()))

  // run
  const output = path.resolve('./test/fixture/output.json')
  const ret = await generate({ preset: 'full', output })

  // verify
  expect(ret).toBe(true)
  expect(mockUtils.writeJSON.mock.calls[0][0]).toEqual(output)
  expect(mockUtils.writeJSON.mock.calls[0][1]).toMatchSnapshot()
})

test('output: fail', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.writeJSON.mockImplementation((path, json) => (Promise.reject(new Error('write error!!'))))

  // run
  const output = path.resolve('./test/fixture/output.json')
  const ret = await generate({ preset: 'default', output })

  // verify
  expect(ret).toBe(false)
  expect(spyError).toHaveBeenCalledWith('write error!!')
})
