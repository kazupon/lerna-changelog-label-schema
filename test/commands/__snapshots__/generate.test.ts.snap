// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output: success 1`] = `
Array [
  Object {
    "aliases": Array [
      "features",
      "enhancement",
    ],
    "color": "ffff00",
    "description": "Includes new features",
    "name": "Type: Feature",
  },
  Object {
    "aliases": Array [
      "bug",
    ],
    "color": "ee0701",
    "description": "Bug or Bug fixes",
    "name": "Type: Bug",
  },
  Object {
    "aliases": Array [
      "vulnerability",
      "vulnerabilities",
    ],
    "color": "b60205",
    "description": "Security fixes",
    "name": "Type: Security",
  },
  Object {
    "aliases": Array [
      "perf",
      "perform",
    ],
    "color": "cc317c",
    "description": "Includes performance fixes",
    "name": "Type: Performance",
  },
  Object {
    "aliases": Array [],
    "color": "1d76db",
    "description": "Includes backwards-compatible fixes",
    "name": "Type: Improvement",
  },
  Object {
    "aliases": Array [
      "breaking",
      "breaking-change",
    ],
    "color": "b60205",
    "description": "Includes backwards-incompatible fixes",
    "name": "Type: Breaking",
  },
  Object {
    "aliases": Array [
      "deprecated",
    ],
    "color": "f7ffa8",
    "description": "Includes deprecate fixes",
    "name": "Type: Deprecated",
  },
  Object {
    "aliases": Array [
      "internationalization",
      "i18n",
    ],
    "color": "ffd412",
    "description": "An internationalization fixes",
    "name": "Type: I18n",
  },
  Object {
    "aliases": Array [
      "accessibility",
      "a11y",
    ],
    "color": "0000ff",
    "description": "An accessibility fixes",
    "name": "Type: A11y",
  },
  Object {
    "aliases": Array [
      "documents",
      "document",
    ],
    "color": "5319e7",
    "description": "A documetation fixes",
    "name": "Type: Documentation",
  },
  Object {
    "aliases": Array [
      "refactor",
      "refactoring",
    ],
    "color": "fbca04",
    "description": "A code change that neither fixes a bug nor adds a feature",
    "name": "Type: Refactoring",
  },
  Object {
    "aliases": Array [
      "test",
      "testing",
    ],
    "color": "0e8a16",
    "description": "Adding missing tests or correcting existing tests",
    "name": "Type: Testing",
  },
  Object {
    "aliases": Array [
      "greenkeeper",
      "maintenance",
    ],
    "color": "e5ef7a",
    "description": "Repository Maintenance",
    "name": "Type: Maintenance",
  },
  Object {
    "aliases": Array [
      "sample",
      "examples",
    ],
    "color": "db0875",
    "description": "Includes example and demo code fixes",
    "name": "Type: Example",
  },
  Object {
    "aliases": Array [
      "dependencies",
    ],
    "color": "ffbce7",
    "description": "Dependencies fixes",
    "name": "Type: Dependency",
  },
  Object {
    "aliases": Array [
      "building",
    ],
    "color": "5319e7",
    "description": "A packaging or a building fixes",
    "name": "Type: Build",
  },
]
`;

exports[`preset: default 1`] = `
"[
  {
    \\"name\\": \\"Type: Feature\\",
    \\"color\\": \\"ffff00\\",
    \\"description\\": \\"Includes new features\\",
    \\"aliases\\": [
      \\"features\\",
      \\"enhancement\\"
    ]
  },
  {
    \\"name\\": \\"Type: Bug\\",
    \\"color\\": \\"ee0701\\",
    \\"description\\": \\"Bug or Bug fixes\\",
    \\"aliases\\": [
      \\"bug\\"
    ]
  },
  {
    \\"name\\": \\"Type: Security\\",
    \\"color\\": \\"b60205\\",
    \\"description\\": \\"Security fixes\\",
    \\"aliases\\": [
      \\"vulnerability\\",
      \\"vulnerabilities\\"
    ]
  },
  {
    \\"name\\": \\"Type: Performance\\",
    \\"color\\": \\"cc317c\\",
    \\"description\\": \\"Includes performance fixes\\",
    \\"aliases\\": [
      \\"perf\\",
      \\"perform\\"
    ]
  },
  {
    \\"name\\": \\"Type: Improvement\\",
    \\"color\\": \\"1d76db\\",
    \\"description\\": \\"Includes backwards-compatible fixes\\",
    \\"aliases\\": []
  },
  {
    \\"name\\": \\"Type: Breaking\\",
    \\"color\\": \\"b60205\\",
    \\"description\\": \\"Includes backwards-incompatible fixes\\",
    \\"aliases\\": [
      \\"breaking\\",
      \\"breaking-change\\"
    ]
  },
  {
    \\"name\\": \\"Type: Deprecated\\",
    \\"color\\": \\"f7ffa8\\",
    \\"description\\": \\"Includes deprecate fixes\\",
    \\"aliases\\": [
      \\"deprecated\\"
    ]
  },
  {
    \\"name\\": \\"Type: I18n\\",
    \\"color\\": \\"ffd412\\",
    \\"description\\": \\"An internationalization fixes\\",
    \\"aliases\\": [
      \\"internationalization\\",
      \\"i18n\\"
    ]
  },
  {
    \\"name\\": \\"Type: A11y\\",
    \\"color\\": \\"0000ff\\",
    \\"description\\": \\"An accessibility fixes\\",
    \\"aliases\\": [
      \\"accessibility\\",
      \\"a11y\\"
    ]
  },
  {
    \\"name\\": \\"Type: Documentation\\",
    \\"color\\": \\"5319e7\\",
    \\"description\\": \\"A documetation fixes\\",
    \\"aliases\\": [
      \\"documents\\",
      \\"document\\"
    ]
  }
]"
`;

exports[`preset: full 1`] = `
"[
  {
    \\"name\\": \\"Type: Feature\\",
    \\"color\\": \\"ffff00\\",
    \\"description\\": \\"Includes new features\\",
    \\"aliases\\": [
      \\"features\\",
      \\"enhancement\\"
    ]
  },
  {
    \\"name\\": \\"Type: Bug\\",
    \\"color\\": \\"ee0701\\",
    \\"description\\": \\"Bug or Bug fixes\\",
    \\"aliases\\": [
      \\"bug\\"
    ]
  },
  {
    \\"name\\": \\"Type: Security\\",
    \\"color\\": \\"b60205\\",
    \\"description\\": \\"Security fixes\\",
    \\"aliases\\": [
      \\"vulnerability\\",
      \\"vulnerabilities\\"
    ]
  },
  {
    \\"name\\": \\"Type: Performance\\",
    \\"color\\": \\"cc317c\\",
    \\"description\\": \\"Includes performance fixes\\",
    \\"aliases\\": [
      \\"perf\\",
      \\"perform\\"
    ]
  },
  {
    \\"name\\": \\"Type: Improvement\\",
    \\"color\\": \\"1d76db\\",
    \\"description\\": \\"Includes backwards-compatible fixes\\",
    \\"aliases\\": []
  },
  {
    \\"name\\": \\"Type: Breaking\\",
    \\"color\\": \\"b60205\\",
    \\"description\\": \\"Includes backwards-incompatible fixes\\",
    \\"aliases\\": [
      \\"breaking\\",
      \\"breaking-change\\"
    ]
  },
  {
    \\"name\\": \\"Type: Deprecated\\",
    \\"color\\": \\"f7ffa8\\",
    \\"description\\": \\"Includes deprecate fixes\\",
    \\"aliases\\": [
      \\"deprecated\\"
    ]
  },
  {
    \\"name\\": \\"Type: I18n\\",
    \\"color\\": \\"ffd412\\",
    \\"description\\": \\"An internationalization fixes\\",
    \\"aliases\\": [
      \\"internationalization\\",
      \\"i18n\\"
    ]
  },
  {
    \\"name\\": \\"Type: A11y\\",
    \\"color\\": \\"0000ff\\",
    \\"description\\": \\"An accessibility fixes\\",
    \\"aliases\\": [
      \\"accessibility\\",
      \\"a11y\\"
    ]
  },
  {
    \\"name\\": \\"Type: Documentation\\",
    \\"color\\": \\"5319e7\\",
    \\"description\\": \\"A documetation fixes\\",
    \\"aliases\\": [
      \\"documents\\",
      \\"document\\"
    ]
  },
  {
    \\"name\\": \\"Type: Refactoring\\",
    \\"color\\": \\"fbca04\\",
    \\"description\\": \\"A code change that neither fixes a bug nor adds a feature\\",
    \\"aliases\\": [
      \\"refactor\\",
      \\"refactoring\\"
    ]
  },
  {
    \\"name\\": \\"Type: Testing\\",
    \\"color\\": \\"0e8a16\\",
    \\"description\\": \\"Adding missing tests or correcting existing tests\\",
    \\"aliases\\": [
      \\"test\\",
      \\"testing\\"
    ]
  },
  {
    \\"name\\": \\"Type: Maintenance\\",
    \\"color\\": \\"e5ef7a\\",
    \\"description\\": \\"Repository Maintenance\\",
    \\"aliases\\": [
      \\"greenkeeper\\",
      \\"maintenance\\"
    ]
  },
  {
    \\"name\\": \\"Type: Example\\",
    \\"color\\": \\"db0875\\",
    \\"description\\": \\"Includes example and demo code fixes\\",
    \\"aliases\\": [
      \\"sample\\",
      \\"examples\\"
    ]
  },
  {
    \\"name\\": \\"Type: Dependency\\",
    \\"color\\": \\"ffbce7\\",
    \\"description\\": \\"Dependencies fixes\\",
    \\"aliases\\": [
      \\"dependencies\\"
    ]
  },
  {
    \\"name\\": \\"Type: Build\\",
    \\"color\\": \\"5319e7\\",
    \\"description\\": \\"A packaging or a building fixes\\",
    \\"aliases\\": [
      \\"building\\"
    ]
  }
]"
`;
